### ISSUE TITLE
QueryMap does not support ConcurrentMap in desugared Android Application

### ISSUE THREAD

**Issue by RavioliWonton** [square/retrofit#3490]
Environment: Retrofit 2.9.0, Kotlin 1.4.20-RC / 1.4.10, Android SDK 30, AGP 4.1.0, coreLibraryDesugaringEnabled true, sourceCompatibility JavaVersion.VERSION_1_8, targetCompatibility JavaVersion.VERSION_1_8, coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.1.0 / 1.0.11'

When I use java.util.concurrent.ConcurrentHashMap in service like this:
```
interface TestInterface {
    @GET("/test")
    suspend fun getTestList(
            @Query("Id") configId: Number,
            @QueryMap paramMap: ConcurrentHashMap<String, String>
    ): Response<List<TestBean>?>
}
```

I will get error like this:

> FATAL EXCEPTION: DefaultDispatcher-worker-1
    Process: com.retrofittest.test_android, PID: 14052
    java.lang.IllegalArgumentException: @QueryMap keys must be of type String: K (parameter #2)
        for method TestInterface.getSelectorList
        at retrofit2.Utils.methodError(Utils.java:54)
        at retrofit2.Utils.methodError(Utils.java:43)
        at retrofit2.Utils.parameterError(Utils.java:64)
        at retrofit2.RequestFactory$Builder.parseParameterAnnotation(RequestFactory.java:499)
        at retrofit2.RequestFactory$Builder.parseParameter(RequestFactory.java:325)
        at retrofit2.RequestFactory$Builder.build(RequestFactory.java:206)
        at retrofit2.RequestFactory.parseAnnotations(RequestFactory.java:67)
        at retrofit2.ServiceMethod.parseAnnotations(ServiceMethod.java:26)
        at retrofit2.Retrofit.loadServiceMethod(Retrofit.java:202)
        at retrofit2.Retrofit$1.invoke(Retrofit.java:160)
        at java.lang.reflect.Proxy.invoke(Proxy.java:393)
        at $Proxy1.getTestList(Unknown Source)

But if I changed ConcurrentHashMap to HashMap, or **turned coreLibraryDesugaringEnabled to off**, the error disappers.

(PS. Please, since we still need support device under API 21, also release the fix to 2.6.x channel. It would be great help!)

**Comment by JakeWharton** [comment]
This is almost certainly a D8/R8 bug where they are not properly rewriting the generic signature of the method in bytecode when changing its type from the built-in CHM to the desugared one.

Can you file a bug at https://issuetracker.google.com/issues/new?component=317603 and link it here?

**Comment by JakeWharton** [comment]
Also providing a complete sample project that reproduces the issue will go a long way to making that guess definitive. If you can send the final APK or a full AS project that I can just build myself I can check the signature in the bytecode. Unfortunately I don't really have the time to set all of the necessary things up to try and recreate it myself.

**Comment by RavioliWonton** [comment]
Sample project [here](https://drive.google.com/file/d/11P8-P6UwVOU-tD6Q8InJtsJyq6899ZvL/view?usp=sharing), and the issue [here](https://issuetracker.google.com/issues/172591375).

**Comment by RavioliWonton** [comment]
Fixed by r8 2.1.80. ([explanation](https://issuetracker.google.com/issues/172591375#comment2))
